protocol/types-and-error-folders.md

```diff
@@ -0,0 +1,23 @@
+# Proposal to move types and errors files to dedicated folders
+
+## Context
+
+The OP Stack contracts in some instances define errors and custom types in a separate file from the contract(s) that use them. This is common among contracts that are part of a common folder e.g L1/, L2/ etc which might have a common error or/and types file where they import the needed errors and types from. This has the advantage of re-usage of errors and types across contracts without duplicating code, ease of modifiability and extensibility and makes it easy to compare errors and types used by different contracts.
+
+## Problem Statement
+
+These error and types files are most times located in a libraries or lib folder e.g `libraries/` `L1/libraries/`, `L2/libraries/` but these are not actual libraries and so do not belong in the libraries folder. This can also make it hard to find at first, especially to a developer newly exploring the codebase.
+
+## Proposed Solution
+
+The proposed solution is to move these error and types files to dedicated folders e.g `L1/errors/`, `L1/types/`, `L2/errors/`, `L2/types/` etc.
+
+## Alternatives Considered
+
+No alternatives considered.
```

Comment:
Regarding my comment above, I think the alternative to consider is that maybe we shouldn't have these type/error files at all. I'm more convinced of this for errors but less convinced of this for types.

Reply:
Yes same here, with #143 i'm okay with errors being defined in the contract file itself. I'm also okay with leaving types as is especially with types that have dedicated libraries for working with them, these are more of libraries and so it should be okay to leave them as is. But re-defining types like structs in each contract bloats the file and makes it harder to navigate.